src/Stratis.Bitcoin.Features.Consensus/Rules/ConsensusRules.cs

```diff
@@ -133,41 +141,68 @@ public void SetErrorHandler(ConsensusRule errorHandler)
         }
 
         /// <inheritdoc />
-        public async Task ValidateAsync(RuleContext ruleContext)
+        public async Task ValidateAndExecuteAsync(RuleContext ruleContext)
         {
-            foreach (var consensusRule in this.consensusRules)
+            await this.ValidateAsync(ruleContext);
+
+            using (new StopwatchDisposable(o => this.PerformanceCounter.AddBlockProcessingTime(o)))
             {
-                var rule = consensusRule.Value;
+                foreach (var ruleDescriptor in this.executionRules)
+                {
+                    await ruleDescriptor.Rule.RunAsync(ruleContext).ConfigureAwait(false);
+                }
+            }
+        }
 
-                if (rule.Attributes.Count == 0 || rule.Attributes.OfType<ValidationRuleAttribute>().Any())
+        /// <inheritdoc />
+        public async Task ValidateAsync(RuleContext ruleContext)
+        {
+            using (new StopwatchDisposable(o => this.PerformanceCounter.AddBlockProcessingTime(o)))
+            {
+                foreach (var ruleDescriptor in this.validationRules)
                 {
-                    await rule.Rule.RunAsync(ruleContext).ConfigureAwait(false);
+                    if (ruleContext.SkipValidation && ruleDescriptor.CanSkipValidation)
+                    {
+                        this.logger.LogTrace("Rule {0} skipped for block at height {1}.", nameof(ruleDescriptor), ruleContext?.BestBlock?.Height);
```

Comment:
`ruleContext?.` => `ruleContext.`

ruleContext can't be null here since it was accessed at the line above

Reply:
ah yeah good spot